File: plugins/agent-skills/skills/claude-delegate/scripts/claude_delegate/pipeline.py
from __future__ import annotations

from pathlib import Path
from typing import Callable

from .common import artifact_paths, write_json
from .contracts import build_delegate_prompt
from .delegate import DelegateRuntime
from .execution_workspace import build_execution_task_packet, prepare_execution_workspace
from .patch_artifact import generate_patch, write_patch_artifact
from .request import build_command_from_request
from .transport import summarize_tool_uses
from .verifier import evaluate_execution_policy, normalize_completion_fields, run_verification
from .workspace import capture_workspace_state, diff_workspace_state


def _should_capture_workspace(request: dict) -> bool:
    policy = request["task_packet"]["execution_policy"]
    observe_roots = policy.get("observe_roots") or []
    return (
        request["assistant_role"] == "implementer"
        or bool(policy.get("allowed_write_paths"))
        or policy.get("max_changed_files") is not None
        or bool(observe_roots)
    )


def _load_parent_handoff(request: dict) -> dict | None:
    lineage = request.get("lineage") or {}
    parent_job_path = lineage.get("parent_job_path")
    if not parent_job_path:
        return None
    from .handoff import load_parent_handoff

    return load_parent_handoff(parent_job_path)


def _request_for_execution(request: dict, workspace) -> dict:
    updated = dict(request)
    updated["execution_workspace"] = workspace.to_dict()
    updated["task_packet"] = build_execution_task_packet(request["task_packet"], workspace)
    updated["prompt"] = build_delegate_prompt(
        updated["task_packet"],
        assistant_role=request["assistant_role"],
        completion_contract=request["completion_contract"],
        delta_prompt=request.get("delta_prompt") or ((request.get("lineage") or {}).get("delta_prompt")),
        parent_handoff=_load_parent_handoff(request),
    )
    updated["command"] = build_command_from_request(updated)
    return updated


def execute_request_pipeline(
    request: dict,
    artifacts_dir: Path,
    *,
    on_spawn: Callable[[int], None] | None = None,
    on_event: Callable[[dict, int], None] | None = None,
) -> dict:
    paths = artifact_paths(artifacts_dir)
    policy = request["task_packet"]["execution_policy"]
    observe_roots = policy.get("observe_roots") or (
        [request["cwd"]] if request["assistant_role"] == "implementer" else []
    )
    exclude_globs = policy.get("exclude_globs") or []
    workspace = prepare_execution_workspace(request)
    request_for_execution = _request_for_execution(request, workspace)

    snapshot_before = None
    envelope: dict | None = None

    try:
        if _should_capture_workspace(request):
            snapshot_before = capture_workspace_state(
                workspace=workspace,
                observe_roots=observe_roots,
                exclude_roots=[str(paths["artifacts_dir"])],
                exclude_globs=exclude_globs,
            )

        runtime = DelegateRuntime(
            request_for_execution,
            artifacts_dir,
            on_spawn=on_spawn,
            on_event=on_event,
        )
        envelope = runtime.execute()

        workspace_changes: list[dict] = []
        if snapshot_before is not None:
            snapshot_after = capture_workspace_state(
                workspace=workspace,
                observe_roots=observe_roots,
                exclude_roots=[str(paths["artifacts_dir"])],
                exclude_globs=exclude_globs,
            )
            workspace_changes = diff_workspace_state(snapshot_before, snapshot_after, request["cwd"])

            # Generate and persist unified diff patch
            patch_content = generate_patch(snapshot_before, snapshot_after, request["cwd"])
            write_patch_artifact(patch_content, paths["patch"])

        envelope = normalize_completion_fields(request, envelope, workspace_changes)
        envelope["task_packet_summary"] = {
            "goal": request["task_packet"]["goal"],
            "task_type": request["task_packet"]["task_type"],
        }

        boundary = evaluate_execution_policy(request, envelope)
        envelope["boundary"] = boundary

        verification = run_verification(
            request,
            envelope,
            execution_cwd=workspace.execution_cwd,
        )
        envelope["verification"] = verification

        if boundary["status"] == "violated":
            envelope["ok"] = False
            envelope["error_type"] = envelope["error_type"] or "boundary_violation"
            envelope["error_message"] = envelope["error_message"] or "; ".join(item["message"] for item in boundary["violations"])

        if verification["status"] == "failed" and verification["fail_on_error"]:
            envelope["ok"] = False
            envelope["error_type"] = envelope["error_type"] or "verification_failed"
            if envelope.get("error_message") is None:
                envelope["error_message"] = "verification commands failed"

        raw_tool_uses = envelope.get("tool_uses", [])
        envelope["tool_use_count"] = len(raw_tool_uses)
        envelope["tool_uses"] = summarize_tool_uses(raw_tool_uses)
    finally:
        workspace.cleanup()

    assert envelope is not None
    envelope["execution_workspace"] = workspace.to_dict()
    write_json(paths["normalized"], envelope)

    # Write durable handoff artifact for cross-job context transfer
    from .handoff import shape_handoff
    handoff = shape_handoff(envelope)
    write_json(paths["handoff"], handoff)

    return envelope

File: plugins/agent-skills/skills/claude-delegate/src/lib/pipeline.ts
import { artifactPaths, writeJson } from "./common.js";
import { buildDelegatePrompt } from "./delegate-prompt.js";
import { executeDelegateRuntime } from "./delegate.js";
import {
  buildExecutionTaskPacket,
  prepareExecutionWorkspace,
  cleanupWorkspace,
  workspaceToDict,
  type ExecutionWorkspace,
} from "./execution-workspace.js";
import { generatePatch, writePatchArtifact } from "./patch-artifact.js";
import { shapeHandoff } from "./handoff.js";
import { loadParentHandoff } from "./handoff.js";
import { summarizeToolUses } from "./transport.js";
import {
  evaluateExecutionPolicy,
  normalizeCompletionFields,
  runVerification,
} from "./verifier.js";
import {
  captureWorkspaceState,
  diffWorkspaceState,
} from "./workspace.js";

function shouldCaptureWorkspace(
  request: Record<string, unknown>,
): boolean {
  const policy = (request.task_packet as Record<string, unknown>)
    .execution_policy as Record<string, unknown>;
  const observeRoots =
    (policy.observe_roots as string[]) ?? [];
  return (
    request.assistant_role === "implementer" ||
    Boolean(
      (policy.allowed_write_paths as string[] | undefined)?.length,
    ) ||
    policy.max_changed_files != null ||
    observeRoots.length > 0
  );
}

function loadParentHandoffForRequest(
  request: Record<string, unknown>,
): Record<string, unknown> | null {
  const lineage = (request.lineage ?? {}) as Record<string, unknown>;
  const parentJobPath = lineage.parent_job_path as string | undefined;
  if (!parentJobPath) return null;
  return loadParentHandoff(parentJobPath);
}

function requestForExecution(
  request: Record<string, unknown>,
  workspace: ExecutionWorkspace,
): Record<string, unknown> {
  const updated = { ...request };
  updated.execution_workspace = workspaceToDict(workspace);
  updated.task_packet = buildExecutionTaskPacket(
    request.task_packet as Record<string, unknown>,
    workspace,
  );
  updated.prompt = buildDelegatePrompt(
    updated.task_packet as Record<string, unknown>,
    {
      assistantRole: request.assistant_role as string,
      completionContract: request.completion_contract as Record<
        string,
        unknown
      >,
      deltaPrompt:
        (request.delta_prompt as string) ??
        ((request.lineage as Record<string, unknown>) ?? {})
          .delta_prompt as string | null ??
        null,
      parentHandoff: loadParentHandoffForRequest(request),
    },
  );
  return updated;
}

export async function executeRequestPipeline(
  request: Record<string, unknown>,
  artifactsDir: string,
  opts?: {
    onSpawn?: ((pid: number) => void) | null;
    onEvent?:
      | ((event: Record<string, unknown>, count: number) => void)
      | null;
  },
): Promise<Record<string, unknown>> {
  const paths = artifactPaths(artifactsDir);
  const policy = (request.task_packet as Record<string, unknown>)
    .execution_policy as Record<string, unknown>;
  const observeRoots = (policy.observe_roots as string[]) ?? (
    request.assistant_role === "implementer"
      ? [request.cwd as string]
      : []
  );
  const excludeGlobs = (policy.exclude_globs as string[]) ?? [];
  const workspace = prepareExecutionWorkspace(request);
  const execRequest = requestForExecution(request, workspace);

  let snapshotBefore: Record<string, Record<string, unknown>> | null = null;
  let envelope: Record<string, unknown> | null = null;

  try {
    if (shouldCaptureWorkspace(request)) {
      snapshotBefore = captureWorkspaceState({
        workspace,
        observeRoots,
        excludeRoots: [paths.artifactsDir],
        excludeGlobs,
      });
    }

    envelope = await executeDelegateRuntime({
      request: execRequest,
      artifactsDir,
      onSpawn: opts?.onSpawn ?? null,
      onEvent: opts?.onEvent ?? null,
    });

    let workspaceChanges: Record<string, unknown>[] = [];
    if (snapshotBefore !== null) {
      const snapshotAfter = captureWorkspaceState({
        workspace,
        observeRoots,
        excludeRoots: [paths.artifactsDir],
        excludeGlobs,
      });
      workspaceChanges = diffWorkspaceState(
        snapshotBefore,
        snapshotAfter,
        request.cwd as string,
      );

      const patchContent = generatePatch(
        snapshotBefore,
        snapshotAfter,
        request.cwd as string,
      );
      writePatchArtifact(patchContent, paths.patch);
    }

    envelope = normalizeCompletionFields(request, envelope, workspaceChanges);
    envelope.task_packet_summary = {
      goal: (request.task_packet as Record<string, unknown>).goal,
      task_type: (request.task_packet as Record<string, unknown>).task_type,
    };

    const boundary = evaluateExecutionPolicy(request, envelope);
    envelope.boundary = boundary;

    const verification = runVerification(
      request,
      envelope,
      { executionCwd: workspace.executionCwd },
    );
    envelope.verification = verification;

    if ((boundary as Record<string, unknown>).status === "violated") {
      envelope.ok = false;
      envelope.error_type = envelope.error_type ?? "boundary_violation";
      envelope.error_message =
        envelope.error_message ??
        (
          (boundary as Record<string, unknown>).violations as Record<
            string,
            unknown
          >[]
        )
          .map((item) => item.message as string)
          .join("; ");
    }

    if (
      (verification as Record<string, unknown>).status === "failed" &&
      (verification as Record<string, unknown>).fail_on_error
    ) {
      envelope.ok = false;
      envelope.error_type =
        envelope.error_type ?? "verification_failed";
      if (envelope.error_message == null) {
        envelope.error_message = "verification commands failed";
      }
    }

    const rawToolUses = (envelope.tool_uses as Record<string, unknown>[]) ?? [];
    envelope.tool_use_count = rawToolUses.length;
    envelope.tool_uses = summarizeToolUses(rawToolUses);
  } finally {
    cleanupWorkspace(workspace);
  }

  envelope!.execution_workspace = workspaceToDict(workspace);
  writeJson(paths.normalized, envelope);

  const handoff = shapeHandoff(envelope!);
  writeJson(paths.handoff, handoff);

  return envelope!;
}
